src/cargo/core/resolver/mod.rs

```diff
@@ -1000,42 +1087,53 @@ fn activate_deps_loop(
-        // cases that we didn't make a backtrack frame in the first place.
-        // This `backtracked` var stores whether we are continuing from a restored backtrack frame
-        // so that we can skip caching `conflicting_activations` in `past_conflicting_activations`
+
+        // When backtracking we don't fully update `conflicting_activations`
+        // especially for the cases that we didn't make a backtrack frame in the
+        // first place.  This `backtracked` var stores whether we are continuing
+        // from a restored backtrack frame so that we can skip caching
+        // `conflicting_activations` in `past_conflicting_activations`
         let mut backtracked = false;
 
-        while !successfully_activated {
-            let next = remaining_candidates.next(cx.prev_active(&dep), &cx.links);
+        loop {
+            let next = remaining_candidates.next(&cx, &dep);
 
-            // Alright, for each candidate that's gotten this far, it meets the
-            // following requirements:
-            //
-            // 1. The version matches the dependency requirement listed for this
-            //    package
-            // 2. There are no activated versions for this package which are
-            //    semver/links-compatible, or there's an activated version which is
-            //    precisely equal to `candidate`.
-            //
-            // This means that we're going to attempt to activate each candidate in
-            // turn. We could possibly fail to activate each candidate, so we try
-            // each one in turn.
             let (candidate, has_another) = next.or_else(|conflicting| {
-                conflicting_activations.extend(conflicting);
-                // This dependency has no valid candidate. Backtrack until we
-                // find a dependency that does have a candidate to try, and try
-                // to activate that one.
+                // If we get here then our `remaining_candidates` was just
+                // exhausted, so `dep` failed to activate.
+                //
+                // It's our job here to backtrack, if possible, and find a
+                // different candidate to activate. If we can't find any
+                // candidates whatsoever then it's time to bail entirely.
                 trace!("{}[{}]>{} -- no candidates", parent.name(), cur, dep.name());
 
+                // Add all the reasons to our frame's list of conflicting
+                // activations, as we may use this to start backtracking later.
+                conflicting_activations.extend(conflicting);
+
+                // Use our list of `conflicting_activations` to add to our
+                // global list of past conflicting activations, effectively
+                // globally poisoning `dep` if `conflicting_activations` ever
+                // shows up again. This is used during backtracking to ensure we
+                // can skip over as much work as possible.
```

Comment:
I miss read `this` as meaning `past_conflicting_activations` instead of `conflicting_activations`